tailfs/shared/readonlydir.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// package shared contains components shared by different tailfs packages
+package shared
+
+import (
+	"errors"
+	"io"
+	"io/fs"
+	"sync"
+)
+
+// DirFile implements webdav.File for a virtual directory.
+// It mimics the behavior of an os.File that is pointing at a real directory.
+type DirFile struct {
+	// Info provides the fs.FileInfo for this directory
+	Info fs.FileInfo
+	// LoadChildren is used to load the fs.FileInfos for this directory's
+	// children. It is called at most once in order to support listing
+	// children.
+	LoadChildren   func() ([]fs.FileInfo, error)
```

Comment:
It's exported because it's specified by the user of DirFile.

What's the benefit of using a constructor, just to make the thing immutable? For simple structs like this I like to avoid the user of constructors because Go doesn't have named method arguments, and as the struct grows, the constructor signature can become unwieldy.